stumpy/search.py

```diff
@@ -0,0 +1,361 @@
+import numpy as np
+
+from . import core
+
+
+def search_k_motifs(
+    T, P, k=1, excl_zone=None, min_neighbors=1, max_occurrences=10, atol=None, rtol=1.0,
+):
+    """Find the top k motifs of the time series `T`.
+
+    A subsequence `Q` is considered a motif if there is at least one other
+    occurrence in `T` (outside the exclusion zone) with distance smaller than
+    `atol + rtol * profile_value`, where `profile_value` is the matrix profile
+    value of `Q`.
+
+    Parameters
+    ----------
+    T : ndarray
+        The time series of interest
+
+    P : ndarray
+        Matrix Profile of `T` (result of a self-join)
+
+    k : int
+        Number of motifs to search. Defaults to `1`.
+
+    excl_zone : int or None
+        Size of the exclusion zone.
+        If `None`, defaults to `m/4`, where `m` is the length of `Q`.
+
+    min_neighbors : int
+        The minimum amount of similar occurrences a subsequence needs to have
+        to be considered a motif.
+        Defaults to `1`. This means, that a subsequence has to have
+        at least one similar occurence to be considered a motif.
+
+    max_occurrences : int
+        The maximum amount of similar occurrences to be returned. The resulting
+        occurrences are sorted by distance, so a value of `10` means that the
+        indices of the most similar `10` subsequences is returned. If `None`,
+        all occurrences are returned.
+        Defaults to `10`
+
+    atol : float or None
+        Absolute tolerance (see equation in description).
+        If `None`, defaults to `0.25 * np.std(P)`
+
+    rtol : float
+        Relative tolerance (see equation in description).
+        Defaults to `1.0`
+
+    Return
+    ------
+    top_motif_indices : list
+        Each item is another list, with each item representing an occurrence of the
+        motif in `T`, sorted by distance from the "original occurrence", i.e. the
+        subsequence of `T` starting at the index in the first place of the list
+
+    top_motif_values : list
+        The matrix profile values of the "original occurrences"
+    """
+    if k < 1:  # pragma: no cover
```

Comment:
Instead of error-ing out, I think it may be better to display a warning:

```
if k < 1:  # pragma: no cover:
    logging.warn("The number of motifs, `k`, must be greater than or equal to 1")
    logging.warn("`k` has been set to `1`")
    k = 1
```

We don't want things to fail if we can do something sensible/reasonable for the user like set it to the default